src/MetalPerformanceShadersGraph/MPSGraphExtensions.cs

```diff
@@ -0,0 +1,60 @@
+#nullable enable
+
+using System;
+using System.Buffers;
+using System.Runtime.InteropServices;
+
+using Foundation;
+using ObjCRuntime;
+using Metal;
+using MetalPerformanceShaders;
+
+namespace MetalPerformanceShadersGraph
+{
+	public static partial class MPSGraphMemoryOps_Extensions
+	{
+		public static unsafe MPSGraphTensor Constant (this MPSGraph graph, float scalar)
+		{
+			return graph.Constant ((double)scalar, new[]{ 1 }, MPSDataType.Float32);
+		}
+
+		public static unsafe MPSGraphTensor Constant (this MPSGraph graph, ReadOnlySpan<float> values, int[] shape)
+		{
+			var length = 1;
+			for (var i = 0; i < shape.Length; i++)
+				length *= shape[i];
+			if (length != values.Length)
+				throw new ArgumentException ($"The number of values ({values.Length}) does not match the shape length ({length}).");
+			fixed (float* p = values) {
+				using var data = NSData.FromBytesNoCopy ((IntPtr)p, (nuint)(values.Length * 4), freeWhenDone: false);
+				return graph.Constant (data, shape, MPSDataType.Float32);
```

Comment:
Is the NSData contents read after the call to Constant returns? If so, NSData might read from a not-fixed-anymore pointer later.

Reply:
I did some tests and as far as I could tell, no, they just copy from it (I checked its retain counts at different times) that's why I can get away with this light-weight function. I've run a fair amount of training with this code and never had memory problems.